正規表現チェッカープログラムの作成

chionyan opened the pull request:
## プログラムの概要

- 起動すると`Text?:`の文言が表示され、正規表現の確認で使うテキストの入力を求められる。
- テキストを入力すると、`Pattern?:`の文言が表示され、正規表現パターンの入力を求められる。
- 正規表現として無効な文字列だった場合は、`Invalidpattern:`の文言に続いて具体的なエラー内容が表示され、再度パターンの入力を求められる。
- 正規表現の入力が終わると、`Matched:`の文言に続いて、すべてのマッチした文字列がカンマ区切りで表示される。
- 1つもマッチしなかった場合は`Nothingmatched.`の文言が表示される。
- マッチング結果を表示したらプログラムを終了する。

## 動作確認

### 1. RSpec
- `$ bundle exec rspec spec/chapter09/regexp_checker_spec.rb` を実行し、下記の出力を確認の上、全てのテストがパスしていることを確認する。（テストの並びは順不同）
```
RegexpChecker
  正規表現として無効な文字列が入力された場合
    エラーが発生すること
  テキストが正規表現にマッチする場合
    "Matched: "に続いて、全てのマッチした文字列が表示されること
  テキストが正規表現にマッチしない場合
    "Nothing matched."が出力されること
…（省略）…
3 examples, 0 failures
```

## その他

- Minitest については無限ループになるコードになっているため、後に修正します。

chionyan committed:
[add]regexp_checker.rb 作成

chionyan committed:
[feature]正規表現チェッカープログラムの作成

- テキストの入力処理を実装
- 正規表現パターンの入力処理を実装
- テキストからパターンに一致するものを String#scan で抜き出す
- 一致するものがあれば一覧を表示、なければその旨を表示
- 正規表現パターン入力の際に RegexpError が発生したら パターン入力を retry するよう実装

chionyan committed:
[add]regexp_check_test, regexp_check_spec を作成

chionyan committed:
[modify]regexp_check のリファクタリング

- spec テスト導入のため、クラス追加
- 各処理におけるメソッドの切り出し

chionyan committed:
[feature]regexp_checker_spec の実装

- before ブロックで RegexpChecker#new を行う
- 標準入力の RegexpChecker#input_text, RegexpChecker#input_pattern のスタブを作成
- 3つのシナリオでのテストを作成

chionyan committed:
[modify]RegexpChecker クラスに YARD 追加

sukechannnn commented on chapter09/regexp_checker.rb:
```diff
@@ -0,0 +1,59 @@
+  # @params text [String] テキスト
+  # @params regexp [Regexp] 正規表現パターン
+  def output(text, regexp)
+    matches = text.scan(regexp)
```
`#output` よりももっとやってることを表す名前にしてほしいです！
ちょっと汎用的な名前すぎるかな。

sukechannnn commented on spec/chapter09/regexp_checker_spec.rb:
```diff
@@ -0,0 +1,33 @@
+    let(:text){ '123-456-789' }
+    let(:pattern){ '[1-6+' }
+    it 'エラーが発生すること' do
+      expect{ @regexp_checker.run }.to raise_error
```
なんのエラーが出るのかもテストできてるとベターです！

chionyan committed:
[modify]RegexpChecker#output をRegexpChecker#output_result にリネーム

chionyan committed:
[feature]regexp_checker_test の実装

chionyan committed:
[modify]RegexpChecker#try_conv_regexp を切り出し、リネーム

chionyan committed:
[modify]エラーが発生する時のテスト修正

- #conv_regexp に #once 追加
- #raise_error の引数に StandardError 追加

chionyan committed:
[fix]テストのエラーの捕捉するクラスをTypeErrorに修正

chionyan commented on spec/chapter09/regexp_checker_spec.rb:
```diff
@@ -28,7 +28,7 @@
     it 'エラーが発生すること' do
       allow(@regexp_checker).to receive(:conv_regexp).once
-      expect{ @regexp_checker.run }.to raise_error(StandardError)
+      expect{ @regexp_checker.run }.to raise_error(TypeError)
```
最初は `RegexpError` 出てるのかと思ってたのですが、
`#try_comv_regexp` の戻り値が `nil` だったため、`#output_result` で捕捉した `TypeError` を引数に取っています。

chionyan replied:
最初は `RegexpError` 出てるのかと思ってたのですが、
`#try_comv_regexp` の戻り値が `nil` だったため、`#output_result` で捕捉した `TypeError` を引数に取っています。

chionyan commented on chapter09/regexp_checker.rb:
```diff
@@ -26,21 +26,21 @@ def input_pattern
   # return [Regexp] 文字列から変換された正規表現クラス
-  def conv_regexp
-    retry_count = 0
+  def conv_regexp(pattern)
```
例外が何が出ているかを具体化する＆[spec テストで呼び出し回数を制御する](https://qiita.com/jnchito/items/640f17e124ab263a54dd#%E3%83%A1%E3%82%BD%E3%83%83%E3%83%89%E3%81%AE%E5%91%BC%E3%81%B3%E5%87%BA%E3%81%97%E5%9B%9E%E6%95%B0%E3%82%92%E6%A4%9C%E8%A8%BC%E3%81%99%E3%82%8B)ために、
`#conv_regexp` を切り出しています。
従来の `#conv_regexp` は `#try_conv_regexp` にリネームしています。

chionyan commented on chapter09/regexp_checker.rb:
```diff
@@ -26,21 +26,21 @@ def input_pattern
+  def try_conv_regexp
     begin
-      Regexp.new(input_pattern)
+      conv_regexp(input_pattern)
```
例外が何が出ているかを具体化する＆[spec テストで呼び出し回数を制御する](https://qiita.com/jnchito/items/640f17e124ab263a54dd#%E3%83%A1%E3%82%BD%E3%83%83%E3%83%89%E3%81%AE%E5%91%BC%E3%81%B3%E5%87%BA%E3%81%97%E5%9B%9E%E6%95%B0%E3%82%92%E6%A4%9C%E8%A8%BC%E3%81%99%E3%82%8B)ために、
`#conv_regexp` を切り出しています。
従来の `#conv_regexp` は `#try_conv_regexp` にリネームしています。

chionyan replied:
例外が何が出ているかを具体化する＆[spec テストで呼び出し回数を制御する](https://qiita.com/jnchito/items/640f17e124ab263a54dd#%E3%83%A1%E3%82%BD%E3%83%83%E3%83%89%E3%81%AE%E5%91%BC%E3%81%B3%E5%87%BA%E3%81%97%E5%9B%9E%E6%95%B0%E3%82%92%E6%A4%9C%E8%A8%BC%E3%81%99%E3%82%8B)ために、
`#conv_regexp` を切り出しています。
従来の `#conv_regexp` は `#try_conv_regexp` にリネームしています。

sukechannnn commented on chapter09/regexp_checker_test.rb:
```diff
@@ -0,0 +1,32 @@
+class RegexpCheckerTest < Minitest::Test
+  def setup
+    @regexp_checker = RegexpChecker.new
+  end
```
よく調べて書いたね👍

sukechannnn commented on spec/chapter09/regexp_checker_spec.rb:
```diff
@@ -0,0 +1,34 @@
+
+RSpec.describe RegexpChecker do
+  before do
+    @regexp_checker = RegexpChecker.new
```
ここは `let!` 使ってほしいです！
RSpec の場合はテストでインスタンス変数はほとんど使わないので、覚えておいてください🙏

chionyan committed:
[clean]regexp_checker_spec コードの整理

- context ブロックでのlet の後に空行追加
- regexp_checker を let! で宣言

chionyan committed:
[modify]#conv_regexp の切り出しを除去

- 切り出していた処理は新しい #conv_regexp の begin式に戻す
- メソッドのリネーム
- regexp_checker_spec のエラー発生時のスタブ対象を Regexp#new に変更

sukechannnn reviewed (approved):
とても良いコードだと思います！
LGTM 👍

daido1976 commented on spec/chapter09/regexp_checker_spec.rb:
```diff
@@ -0,0 +1,38 @@
+require './chapter09/regexp_checker'
+
+RSpec.describe RegexpChecker do
+  let!(:regexp_checker){ RegexpChecker.new }
```
ここは以下のように `subject` に置き換えられそうです！
```ruby
subject { RegexpChecker.new.run }
```
